### DISCUSSION TITLE
CGLIB에 대한 설명과 동작방식

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#185] [Spring Core]
.

**Comment by gyungchan-jo** [comment]
CGLIB(Code Generation Library)은 Java의 런타임 시에 클래스의 프록시를 생성하기 위해 사용되는 강력한 라이브러리입니다. CGLIB은 주로 동적 프록시를 생성하는 데 사용되며, Spring Framework에서도 AOP(Aspect-Oriented Programming) 기능을 구현할 때 널리 사용됩니다.

CGLIB의 주요 특징
- 바이트코드 조작: CGLIB은 바이트코드 조작을 통해 런타임 시에 새로운 클래스를 동적으로 생성합니다.
- 메소드 오버라이딩: CGLIB은 특정 클래스의 메소드를 오버라이드하여 새로운 동작을 추가할 수 있습니다.
- 인터페이스 필요 없음: Java의 기본 프록시(Proxy) 클래스와 달리, CGLIB은 인터페이스가 없더라도 클래스를 프록시할 수 있습니다.
- 성능: CGLIB은 런타임 시에 바이트코드를 생성하므로, 실행 성능이 비교적 뛰어납니다.

CGLIB의 동작 방식

1. 클래스 확장: CGLIB은 프록시하려는 클래스를 확장하여 새로운 서브 클래스를 생성합니다.
2. 메소드 오버라이딩: 생성된 서브 클래스는 원본 클래스의 메소드를 오버라이드하여 필요한 추가 동작을 구현합니다.
3. 바이트코드 생성: 런타임 시에 ASM 라이브러리를 사용하여 바이트코드를 생성하고, 이를 통해 새로운 클래스를 메모리에 로드합니다.
4. 프록시 인스턴스 반환: 생성된 서브 클래스의 인스턴스를 반환하여 프록시 객체로 사용합니다.

**Comment by nayonsoso** [comment]
### CGLIB 란

- Code Generator Library
- 클래스의 바이트코드를 조작하여 Proxy 객체를 생성해주는 라이브러리이다.
- spring AOP에서는 런타임에 동적으로 자바 클래스의 프록시를 생성하는데, 
  - 타깃이 하나 이상의 인터페이스를 구현한다면 JDK Dynamic Proxy를
  - 타깃이 어떤 인터페이스도 구현하지 않는다면 CGLIB를 사용한다.

cf. JDK Dynamic Proxy란, Java의 리플렉션 패키지에 존재하는 Proxy라는 클래스를 통해 동적으로 생성된 프록시 객체를 의미한다.

### CGLIB 특징
- 인터페이스 없이 단순 클래스만으로 프록시 객체를 동적으로 생성이 가능하다.
- 바이트 코드를 직접 조작하기 때문에 리플렉션을 사용하는 JDK Dynamic Proxy에 비해 성능이 좋다.
- Spring core 패키지에 포함되어있으므로 spring을 사용한다면 별도의 의존성 추가는 필요하지 않다.

### CGLIB 작동 원리
- CGLIB는 타깃 클래스를 상속하고 모든 메소드를 재정의해서 프록시 객체를 생성한다.
- 따라서 CGLIB로 프록시 객체를 만들 때 타깃에 private, final 메서드가 있으면 안된다.
- 또한 반드시 기본 생성자를 만들어주어야 한다.


ref : 
https://huisam.tistory.com/entry/springAOP
https://youtu.be/MFckVKrJLRQ?si=Mu94fxY1yPZD4Hu0 (테코톡)
